malariagen_data/anoph/hap_freq.py

```diff
@@ -0,0 +1,391 @@
+            freq_cols["frq_" + coh] = list(hap_dict.values())
+
+        df_freqs = pd.DataFrame(freq_cols, index=hap_dict.keys())
+
+        # Compute max_af.
+        df_max_af = pd.DataFrame({"max_af": df_freqs.max(axis=1)})
+
+        # Build the final dataframe.
+        df_haps = pd.concat([df_freqs, df_max_af], axis=1)
+
+        df_haps_sorted = df_haps.sort_values(by=["max_af"], ascending=False)
+        df_haps_sorted["label"] = ["H" + str(i) for i in range(len(df_haps_sorted))]
+
+        # Reset index after filtering.
+        df_haps_sorted.set_index(keys="label", drop=True)
+
+        return df_haps_sorted
+
+    @check_types
+    @doc(
+        summary="""
+            Group samples by taxon, area (space) and period (time), then compute
+            haplotype frequencies.
+        """,
+        returns="""
+            The resulting dataset contains data has dimensions "cohorts" and
+            "variants". Variables prefixed with "cohort" are 1-dimensional
+            arrays with data about the cohorts, such as the area, period, taxon
+            and cohort size. Variables prefixed with "variant" are
+            1-dimensional arrays with data about the variants, such as the
+            contig, position, reference and alternate alleles. Variables
+            prefixed with "event" are 2-dimensional arrays with the allele
+            counts and frequency calculations.
+        """,
+    )
+    def haplotypes_frequencies_advanced(
+        self,
+        region: base_params.region,
+        area_by: frq_params.area_by,
+        period_by: frq_params.period_by,
+        sample_sets: Optional[base_params.sample_sets] = None,
+        sample_query: Optional[base_params.sample_query] = None,
+        min_cohort_size: base_params.min_cohort_size = 10,
+        ci_method: Optional[frq_params.ci_method] = frq_params.ci_method_default,
+        chunks: base_params.chunks = base_params.native_chunks,
+        inline_array: base_params.inline_array = base_params.inline_array_default,
+    ) -> xr.Dataset:
+        # Load sample metadata.
+        df_samples = self.sample_metadata(
+            sample_sets=sample_sets, sample_query=sample_query
+        )
+
+        # Prepare sample metadata for cohort grouping.
+        df_samples = _prep_samples_for_cohort_grouping(
+            df_samples=df_samples,
+            area_by=area_by,
+            period_by=period_by,
+        )
+
+        # Group samples to make cohorts.
+        group_samples_by_cohort = df_samples.groupby(["taxon", "area", "period"])
+
+        # Build cohorts dataframe.
+        df_cohorts = _build_cohorts_from_sample_grouping(
+            group_samples_by_cohort=group_samples_by_cohort,
+            min_cohort_size=min_cohort_size,
+        )
+
+        # Early check for no cohorts.
+        if len(df_cohorts) == 0:
+            raise ValueError(
+                "No cohorts available for the given sample selection parameters and minimum cohort size."
+            )
```

Comment:
Feels like this could be moved inside `_build_cohorts_from_sample_grouping()` to reduce code duplication.